src/provider/provider_tracking.c

```diff
@@ -85,16 +203,34 @@ static umf_result_t umfMemoryTrackerRemove(umf_memory_tracker_handle_t hTracker,
     // Every umfMemoryTrackerAdd(..., ptr, ...) should have a corresponding
     // umfMemoryTrackerRemove call with the same ptr value.
 
-    void *value = critnib_remove(hTracker->alloc_segments_map, (uintptr_t)ptr);
+    tracker_alloc_info_t *parent_value = NULL;
+    uintptr_t parent_key = 0;
+    int level = 0;
+
+    // Find the most nested (on the highest level) entry in the map
+    // with the `ptr` key and with no children - only such entry can be removed.
+    tracker_alloc_info_t *value = map_get_most_nested_entry(
```

Comment:
What if the client's code is malformed and it deallocates segment from level N, but there is a segment on level N+1? According to the current logic this code will remove segment from the level N+1, am I right?

Reply:
Yes, but there is a logic in `trackingFree()` that if `umfMemoryProviderFree()` fails (because of wrong size), then the removed segment will be re-added.

Reply:
Could you please add such a test with the Fixed provider?
E.g.

1. Create some pool A
2. Allocated buffer B from pool A.
3. Use buffer B as an input parameter for the Fixed provider instance.
4. Create Pool C with the Fixed provider and allocate buffer D from pool C.
5. Try to free buffer B from pool A and verify how UMF behaves.

Reply:
OK, I will add such test.

Reply:
I have added the `TEST_P(TrackingProviderTest, reverted_free_half_size)` test to https://github.com/oneapi-src/unified-memory-framework/pull/1144

Reply:
> What if the client's code is malformed and it deallocates segment from level N, but there is a segment on level N+1? According to the current logic this code will remove segment from the level N+1, am I right?

Yes, but only in case the sizes are the same - then we cannot distinguish them (see the `TEST_P(TrackingProviderTest, reverted_free_the_same_size)`  test). When the sizes are different it will not happen (see the `TEST_P(TrackingProviderTest, reverted_free_half_size)` test).

Reply:
@vinser52 I added the tests.

Reply:
I added a comment to the test itself.

Reply:
OK, so can you resolve this issue?